### ISSUE TITLE
Remove special-case casting support for numpy<2.0

### ISSUE THREAD

**Issue by greglucas** [pyproj4/pyproj#1309]
Once we support only Numpy 2.0+ we can remove the casting checks we have added to the Cython code for single-point transforms and Geod's fwd/inv single-point as well. Right now we make sure to throw a TypeError so we get the right conversions, but Numpy 2.0+ will do this for us automatically.

https://github.com/pyproj4/pyproj/blob/f235f9b3067e4817ec34a95cfd1717d58fc95293/pyproj/_transformer.pyx#L768-L770

**Comment by snowman2** [comment]
Possibly helpful guide to follow on numpy support window: https://scientific-python.org/specs/spec-0000/

**Comment by djhoese** [comment]
So could this work be done for the September 2025 release (3.8.0) of pyproj?

**Comment by snowman2** [comment]
> So could this work be done for the September 2025 release (3.8.0) of pyproj?

It appears that it is time according to the schedule of support. I would support making the change.

**Comment by djhoese** [comment]
@greglucas Is the correct fix here to remove all `if isinstance: TypeError` lines from:

https://github.com/pyproj4/pyproj/blob/878615decadbde676f9acc82bce9763d62fbacbe/pyproj/_transformer.pyx#L750-L764

So the TODO to resolve this issue is:

1. Remove the above mentioned lines
2. Remove the comment about type-checking
3. Set 2.0 limit for numpy in dependencies in pyproject.toml
4. Optional: Set C Macros for a target numpy version and/or the numpy [deprecated api macro](https://numpy.org/doc/stable/reference/c-api/deprecations.html)

Right?

**Comment by greglucas** [comment]
@djhoese, thanks for taking this on! It has been a while since I touched that, but I think I agree with that list of items to update. Make sure to do it for the geod fwd/inv functions too. Ping me for a review on any PRs and I can try and test with Cartopy too to make sure we are getting the behaviors we were looking for over there where I was trying to optimize the single-point transforms for that project.

**Comment by djhoese** [comment]
Thanks @greglucas. Rewording this, if we remove these checks, we expect numpy to raise a TypeError when trying to use the array as a scalar, right? Would it maybe be "cleaner" to just check the scalar-ness of the data instead of a `try/except TypeError`?

**Comment by greglucas** [comment]
>  if we remove these checks, we expect numpy to raise a TypeError when trying to use the array as a scalar, right? 

I believe that is correct in numpy 2.0, but testing it now I'm not so sure... I still only get a Deprecation warning saying "in the future".

```
>>> float(np.array([1.]))
<stdin>:1: DeprecationWarning: Conversion of an array with ndim > 0 to a scalar is deprecated, and will error in future. Ensure you extract a single element from your array before performing this operation. (Deprecated NumPy 1.25.)
```

> Would it maybe be "cleaner" to just check the scalar-ness of the data instead of a try/except TypeError?

I think this had to do with performance. The try/except was faster, but I'm sure we didn't test all combinations of options for narrowing the types allowed.

Here is the PR https://github.com/pyproj4/pyproj/pull/1204
with a few comments/discussions embedded within for historical reference.
https://github.com/pyproj4/pyproj/pull/1204/files#r1052895488

**Comment by snowman2** [comment]
> Set 2.0 limit for numpy in dependencies in pyproject.toml

Numpy isn't a dependency of pyproj. Will need lots of warnings in the documentation instead.

**Comment by djhoese** [comment]
Well...shoot. That does complicate things doesn't it.

**Comment by snowman2** [comment]
I am thinking this may be better to do on a major release due to the backwards incompatible breaking nature of this change. May also not be bad to give it another year unless there is a pressing need.

**Comment by djhoese** [comment]
Maybe a review of upcoming expected changes and split them up in one or two more releases? There's the PROJ 9.7.0 release, then what's next? It might be nice to have the PR ready and rebase it repeatedly to make sure it functions as expected over time then after the last non-major release we merge it and let it sit to be tested. Oh or a long release candidate or beta on pypi? My other packages tend to have an unstable CI environment that should pull pyproj in from github I think or maybe PyPI prerelease.

**Comment by snowman2** [comment]
I think a long release candidate makes sense. All good suggestions.

**Comment by djhoese** [comment]
Yeah I'm mostly worried about needing to do a maintenance branch for the last non-major release which I always find annoying.

**Comment by snowman2** [comment]
Could do it the other way around and keep the main branch as the maintenance branch and a separate branch for the next major release with a running PR to the main branch as you suggested.

**Comment by NMaiyu** [comment]
Hello, 
I got a DeprecationWarning while using another package (cartopy) that calls pyproj's Geod inv and fwd functions, as it sends 2-dimensional numpy arrays even when it only contains a scalar. 
From what I was able to see, the DeprecationWarning is raised while trying to convert the array to a float in `_copytobuffer_return_scalar`. 
I was able to fix the issue for me by adding `xxx = xxx.flatten()` just before `_copytobuffer_return_scalar` is called by `_copytobuffer` (L128 in pyproj/utils.py), although this might not be suitable in the general case ? 
https://github.com/pyproj4/pyproj/blob/13ca2ab5b97280a012ba47ca482fd2a54a1a4448/pyproj/utils.py#L123-L129

**Comment by djhoese** [comment]
@NMaiyu Thanks for mentioning this case, but I'm a little confused. You mentioned that cartopy is sending 2-dimensional arrays with a single scalar inside, but this `if` statement here is specifically checking for an empty shape (`xxx.shape == ()`) which should mean a scalar array (no dimensions). Could you show us a minimal example of a numpy array that matches what you're seeing?

**Comment by NMaiyu** [comment]
@djhoese Thank you for your answer, when trying to make the example I figured out that I had some old code deactivating filters on warnings hence why I was seeing the DeprecationWarning (although the reason why the 'flatten' had any effect is unclear). 
I am sorry for my erroneous comment.

**Comment by greglucas** [comment]
I just put up a PR to remove the try/except and warning case and also improve the performance. I tried it with Cartopy and all looked good over there still and improved things too.
